docs/proposals/cluster-capacity.md

```diff
@@ -0,0 +1,461 @@
+Some of the volumes can be claimed and provisioned with volume dynamic provisioner.
+Some pods are meant to be run on dedicated nodes.
+
+In general, requirements can include:
+
+* cpu
+* memory
+* required disk space (a.k.a disk bytes)
+* minimal disk IOPs
+* gpu cores
+* volumes
+* zones/nodes specified by labels, taints
+
+Resources can be divided between first-class resources (memory, cpu, disk, etc.) and non-first class resources.
+
+Other resource types can include (e.g. see [4]):
+
+* Hadoop slots
+* Network bandwith (a.k.a NIC b/w)
+* Network operations
+* Network packates per second
+* Storage space
+* Storage time
+* Storage operations
+
+### Use cases
+
+**Use Case 1**: A namespace is configured to allow it to allocate 40GB of memory.
+The POD with the largest amount of memory limit is 4GB.
+The namespace has a node label selector for `region1, compute nodes`.
+We need to know how many 4GB pods can be placed given the physical resources available
+on the nodes that match the node selector of the namespace.
+We also need to know how many of the largest pod definitions can be scaled up on the
+namespace (in this case, the 4GB pod can be scaled up to 10,
+assuming it's the only one defined).
+
+**Use Case 2**: A namespace is configured to allow for 4 total CPUs to be used.
+The POD with the largest amount of CPU allocated is given half a CPU.
+The namespace has a node label selector for `region2, compute nodes`.
+We need to know how many half CPU pods can be placed given the physical resources available
+on the nodes that match the node selector of the namespace.
+We also need to know how many of the largest pod definitions can be scaled up
+on this namespace (in this case, the half CPU pod can be scaled up to 8,
+assuming it's the only one defined).
+
+**Use Case 3**: A POD definition has a PV claim.
+We need to know how many nodes have available attachments such that
+the POD can be scheduled to nodes that fit the namespace node selector.
+
+### Autoscaling
+
+Introduction of new nodes or increase of node resources is not the only way how to deal with insufficient cluster capacity.
+One can take advantage of Horizontal Pod Autoscaling [1] scaling a number of replicas (Replication Controller, Deployment, ReplicaSet) based on the current resource utilization.
+Or Cluster Autoscaler [1].
+Still, the autoscaling is another indicator of insufficient resources which can result in node scaling.
+
+## Design considerations and design space
+
+Assuming the following:
+
+* predicting remaining cluster capacity on pod bases (e.g. "How many pods of a given shape I can schedule")
+* using the scheduler workflow as it is (applying the same predicates and priority functions/executing the same code wherever possible, using the same configuration)
+* prediction must not change the state of the cluster
+* provide general framework usable for any scheduling algorithm
+
+Scheduling process uses two sets of functions: predicates and priority functions.
```

Comment:
I could transform the paragraph into an example. "Some schedulers search the most suitable node based on ...". The paragraph is here to say the scheduler is not a closed system. It may require to query the cluster for additional data which are not part of pods or nodes.